pkg/client/unit.go

```diff
@@ -264,7 +266,7 @@ func (u *Unit) updateState(exp UnitState, logLevel UnitLogLevel, cfg *proto.Unit
 	}
 	if u.configIdx != cfgIdx {
 		u.configIdx = cfgIdx
-		if u.config != cfg {
+		if !reflect.DeepEqual(u.config.Source.AsMap(), cfg.Source.AsMap()) {
```

Comment:
Since `config.Source` is a a proto message of type `google.protobuf.Struct` you should compare them using https://pkg.go.dev/google.golang.org/protobuf/proto#Equal

Comparing as maps sort of works around this but I am not sure how reliable it is. Each instance of a protobuf message contains several internal fields (serialization caches, lists of unparseable fields, etc) that aren't necessarily stable that the comparison needs to know to ignore

Reply:
This is probably causing the unexpected index updates. Not being able to compare proto messages in the obvious ways is one of the primary annoyances of proto messages in Go unfortunately.

Usually the best way to detect these problems is to do a comparison, then marshal one of the messages to populate the `sizeCache`, then compare the messages again and it will fail unless you used `proto.Equal`.

Reply:
> Comparing as maps sort of works around this but I am not sure how reliable it is. Each instance of a protobuf message contains several internal fields (serialization caches, lists of unparseable fields, etc) that aren't necessarily stable that the comparison needs to know to ignore

Yah, that's why I did `Souce.AsMap()`, since that'll just return the public, exported fields, and not any of the internal protobuf struct fields.  Didn't know the protobuf library has its own `Equal` function, should probably switch to that though.

Reply:
Yah, agreed that `poto.Equal` is definitely a more sound choice here.

And yah, after I update this PR I can open a similar once against elastic-agent.

Reply:
You also need the client version bump here updated in Beats, Fleet Server, and possibly APM server in addition to Agent.